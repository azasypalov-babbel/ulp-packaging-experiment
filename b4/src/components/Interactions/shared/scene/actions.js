import * as types from './types';

export const requestPrompt = (attemptText) => ({
  type: types.REQUEST_PROMPT,
  payload: {
    attemptText
  }
});

export const receivePrompt = (attempt) => ({
  type: types.RECEIVE_PROMPT,
  payload: {
    ...attempt
  }
});

export const play = () => ({
  type: types.PLAY
});
