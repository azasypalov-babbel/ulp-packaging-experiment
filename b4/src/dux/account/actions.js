import * as types from './types';
import services from '../../services';

export const fetchAccount = () => (dispatch, getState) => {
  const { accountService } = services;
  const { locale } = getState().session;

  return dispatch({
    type: types.FETCH_ACCOUNT,
    payload: accountService.fetchAccount(locale)
  });
};
