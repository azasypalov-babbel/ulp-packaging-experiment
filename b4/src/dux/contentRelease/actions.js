import * as types from './types';
import services from '../../services';
import rollbar from '../../services/rollbarService';

export const fetchContentRelease = () => (dispatch, getState) => {
  const { contentReleaseService } = services;
  const { locale, learnLanguageAlpha3 } = getState().session;

  rollbar.warning('Fetch ContentReleaseId - not provided by the lesson url');

  return dispatch({
    type: types.FETCH_CONTENT_RELEASE,
    payload: contentReleaseService.fetchContentRelease({ locale, learnLanguageAlpha3 })
  });
};

export const setContentReleaseId = (id) => ({
  type: types.SET_CONTENT_RELEASE_ID,
  payload: id
});
