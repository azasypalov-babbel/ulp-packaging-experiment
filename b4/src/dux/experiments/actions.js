import * as types from './types';
import services from '../../services';
import camelize from 'camelize';

export const fetchExperimentBucket = (target) => (dispatch, getState) => {
  const { abTesterService } = services;
  const { locale, learnLanguageAlpha3 } = getState().session;

  return dispatch({
    type: types.FETCH_EXPERIMENT_BUCKET,
    payload: abTesterService.getExperimentBucket({ locale, learnLanguageAlpha3, target })
      .then((variation) => ({
        target: camelize(target),
        variation
      }))
      .catch((err) => {
        throw { ...err, target: camelize(target) };
      })
  });
};
