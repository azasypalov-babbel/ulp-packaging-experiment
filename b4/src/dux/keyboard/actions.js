import * as types from './types';

export const toggleHints = () => ({
  type: types.TOGGLE_HINTS
});
