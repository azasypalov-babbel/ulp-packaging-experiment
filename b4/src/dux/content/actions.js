import * as types from './types';
import services from '../../services';
import * as contentReleaseSelectors from '../contentRelease/selectors';

export const fetchLearnLanguage = () => (dispatch, getState) => {
  const { contentService } = services;
  const { contentRelease, session } = getState();
  const { locale, learnLanguageAlpha3 } = session;
  const contentReleaseId = contentReleaseSelectors.contentReleaseId(contentRelease);

  return dispatch({
    type: types.FETCH_LEARN_LANGUAGE,
    payload: contentService.fetchLearnLanguage({
      locale,
      contentReleaseId,
      learnLanguageAlpha3
    })
  });
};
