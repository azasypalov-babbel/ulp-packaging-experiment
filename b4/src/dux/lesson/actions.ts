import * as types from './types';
import services from '../../services';
import { navigateToReturnUrl } from '../session/actions';
import * as lessonSessionEvents from '../tracker/events/lessonSession';
import * as contentSelectors from '../content/selectors';
import * as contentReleaseSelectors from '../contentRelease/selectors';
import * as features from '../../lib/features';
import { track } from '../tracker/actions';
import { abortSequence } from '../sequence/actions';
import { AppThunk, RootState } from '@src/store';
import { IGetLessonParams } from '@src/services/lesson/interface';

const lessonPayload = (getState: () => RootState) => {
  const { contentRelease, lesson, session, content } = getState();
  const { locale, learnLanguageAlpha3 } = session;
  const { lessonUuid, learningActivityId } = lesson;
  const { id: courseId } = contentSelectors.currentCourse(
    content,
    learningActivityId
  );
  const { id: courseOverviewId } = contentSelectors.currentCourseOverview(
    content,
    learningActivityId
  );
  const contentReleaseId = contentReleaseSelectors.contentReleaseId(contentRelease);

  return {
    contentReleaseId,
    learnLanguageAlpha3,
    locale,
    lessonUuid,
    learningActivityId,
    courseId,
    courseOverviewId
  };
};

export const initLesson = ({ contentReleaseId, lessonUuid, learningActivityId }: Partial<IGetLessonParams>) => {
  return {
    type: types.INIT_LESSON,
    payload: {
      contentReleaseId,
      lessonUuid,
      learningActivityId
    }
  };
};

export const fetchLessonData = (): AppThunk => (dispatch, getState) => {
  const { lessonService } = services;
  const { contentRelease, lesson, session } = getState();
  const { locale, learnLanguageAlpha3 } = session;
  const { lessonUuid, learningActivityId } = lesson;
  const contentReleaseId = contentReleaseSelectors.contentReleaseId(contentRelease);

  return dispatch({
    type: types.FETCH_LESSON_DATA,
    payload: lessonService.getLessonData({
      contentReleaseId,
      locale,
      learnLanguageAlpha3,
      lessonUuid,
      learningActivityId
    })
  });
};

export const postLessonCompleted = (): AppThunk => (dispatch, getState) => {
  const { lessonService } = services;

  return dispatch({
    type: types.POST_LESSON_COMPLETE,
    payload: lessonService.postLessonCompleted(lessonPayload(getState))
  });
};

export const completeLesson = (): AppThunk => (dispatch) => {
  return dispatch({
    type: types.COMPLETE_LESSON,
    payload: dispatch(postLessonCompleted())
  });
};

export const closeLesson = (): AppThunk => async (dispatch, getState) => {
  dispatch({
    type: types.CLOSE_LESSON
  });

  const { sequence: { completed: isLessonCompleted } } = getState();

  if (features.isWebview()) {
    const { lessonService } = services;
    if (!isLessonCompleted) {
      dispatch(abortSequence());
      lessonService.postLessonAbort(lessonPayload(getState));
    } else {
      lessonService.closeLesson();
    }
  } else {
    if (!isLessonCompleted) {
      // For web, abortSequence happens in unload handler in TrainersSequence;
    }
    dispatch(navigateToReturnUrl());
  }
};

export const preparePurge = (): AppThunk => (dispatch, getState) => {
  const event = lessonSessionEvents.lessonSessionStartedPurgeEvent(getState());
  return dispatch(track(event));
};
