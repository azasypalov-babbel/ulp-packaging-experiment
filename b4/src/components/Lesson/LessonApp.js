import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import PropTypes from 'prop-types';

import * as trackerActions from '../../dux/tracker/actions';
import * as lessonActions from '../../dux/lesson/actions';
import * as statisticsActions from '../../dux/statistics/actions';
import * as profileActions from '../../dux/profile/actions';
import * as subscriptionsActions from '../../dux/subscriptions/actions';
import * as accountActions from '../../dux/account/actions';

import * as statisticsSelectors from '../../dux/statistics/selectors';
import * as profileSelectors from '../../dux/profile/selectors';

import { isAllowedReferAFriend } from '../../lib/smartSurfaces';
import rollbar from '../../services/rollbarService';

import ApplicationLayout from '../shared/ApplicationLayout';
import NavbarContainer from '../shared/Navbar/NavbarContainer';
import FeatureToggles from '../shared/FeatureToggles';
import LessonSequence from './LessonSequence';
import MessagesContainer from '../shared/Messages/MessagesContainer';
import { BottomLayout } from '../shared/BottomLayout';
import ApplicationErrorBoundary from '../ApplicationErrorBoundary';
import TrainerOverlay from '../shared/TrainerOverlay';
import { performancePageLoadedEvent } from '../../dux/tracker/events/performance';

const userHasNoCompletedLessons = async (dispatch) => {
  const { value: statistics } = await dispatch(statisticsActions.fetchTrainerItemsStatistics());
  return !statisticsSelectors.userHasCompletedLessons({ data: statistics });
};

const userIsB2b = async (dispatch) => {
  const { value: profile } = await dispatch(profileActions.fetchProfile());
  return profileSelectors.isB2bUser({ data: profile });
};

const mapStateToProps = (state) => ({
  state,
  locale: state.session.locale,
  learnLanguageAlpha3: state.session.learnLanguageAlpha3
});

export const fetchSmartSurfacesData = () => async (dispatch) => {
  try {
    if (await userHasNoCompletedLessons(dispatch) || await userIsB2b(dispatch)) return;

    return dispatch(subscriptionsActions.fetchSubscriptions());
  } catch (error) {
    rollbar.warning(`SmartSurfacesError@LessonApp: ${error.message}`, {
      message: error.message,
      stack: error.stack
    });
  }
};

const mapDispatchToProps = {
  initLesson: lessonActions.initLesson,
  fetchAccount: accountActions.fetchAccount,
  track: trackerActions.track,
  fetchSmartSurfacesData: fetchSmartSurfacesData
};

export class LessonApp extends React.Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.initLesson();
  }
  async initLesson() {
    const {
      fetchAccount,
      initLesson,
      fetchSmartSurfacesData,
      track,
      locale,
      learnLanguageAlpha3
    } = this.props;

    const { contentReleaseId, lessonUuid, learningActivityId } = this.props;
    initLesson({ contentReleaseId, lessonUuid, learningActivityId });

    await fetchAccount();

    /* eslint-disable camelcase */
    track({
      event: 'product:view_shown',
      version: 1,
      payload: {
        view_name: 'lesson',
        locale: locale,
        learn_language_alpha3: learnLanguageAlpha3
      }
    });
    /* eslint-enable camelcase */

    track(performancePageLoadedEvent({ locale, learnLanguageAlpha3 }));

    // Although this feature is used at the end of the lesson, the relevant data is fetched here so as not
    // to delay the rendering of the LessonEndScreen
    if (isAllowedReferAFriend()) {
      fetchSmartSurfacesData();
    }
  }

  render() {
    return (
      <ApplicationLayout>
        <ApplicationErrorBoundary name="LessonApp">
          <NavbarContainer />
          <FeatureToggles />
          <MessagesContainer />
          <TrainerOverlay />
          <main>
            <LessonSequence />
          </main>
          <BottomLayout />
        </ApplicationErrorBoundary>
      </ApplicationLayout>
    );
  }
}

LessonApp.propTypes = {
  initLesson: PropTypes.func.isRequired,
  fetchAccount: PropTypes.func.isRequired,
  track: PropTypes.func.isRequired,
  fetchSmartSurfacesData: PropTypes.func.isRequired,
  lessonUuid: PropTypes.string.isRequired,
  locale: PropTypes.string.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  contentReleaseId: PropTypes.string,
  learningActivityId: PropTypes.string
};

export default compose(
  connect(mapStateToProps, mapDispatchToProps)
)(LessonApp);
