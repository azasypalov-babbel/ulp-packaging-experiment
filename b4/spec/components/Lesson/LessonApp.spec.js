import React from 'react';
import { shallow } from 'enzyme';

import mockStore from '../../dux/mockStore';

jest.mock('../../../src/components/Lesson/LessonSequence', () => () => null);

jest.mock('../../../src/lib/smartSurfaces');
import { isAllowedReferAFriend } from '../../../src/lib/smartSurfaces';

jest.mock('../../../src/dux/statistics/actions');
import { fetchTrainerItemsStatistics } from '../../../src/dux/statistics/actions';

jest.mock('../../../src/dux/profile/actions');
import { fetchProfile } from '../../../src/dux/profile/actions';

jest.mock('../../../src/dux/subscriptions/actions');
import { fetchSubscriptions } from '../../../src/dux/subscriptions/actions';

jest.mock('../../../src/services/rollbarService');
import rollbar from '../../../src/services/rollbarService';

jest.mock('../../../src/dux/tracker/events/lessonSession', () => ({
  lessonSessionEndedAbortEvent: () => ({ event: 'lesson_session:ended:abort' })
}));

import { fetchSmartSurfacesData, LessonApp } from '../../../src/components/Lesson/LessonApp';

const storeMock = {
  session: {
    locale: 'en',
    learnLanguageAlpha3: 'DEU'
  }
};

const defaultProps = {
  initLesson: jest.fn(),
  setContentReleaseId: jest.fn(),
  track: jest.fn(),
  fetchSmartSurfacesData: jest.fn(),
  fetchAccount: jest.fn(),
  locale: 'en',
  learnLanguageAlpha3: 'DEU',
  fetchExperimentBucket: jest.fn().mockReturnValue(Promise.resolve()),
  lessonUuid: 'mock-lesson-uuid',
  learningActivityId: 'mock-learning-activity-id',
  contentReleaseId: 'mock-content-release-id'
};

describe('LessonApp', () => {
  describe('default behaviour', () => {
    let wrapper;
    beforeEach(() => {
      isAllowedReferAFriend.mockReturnValue(true);
      wrapper = shallow(<LessonApp {...defaultProps} />);
    });

    it('should render', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should initialise lesson', () => {
      expect(defaultProps.initLesson)
        .toHaveBeenCalledWith({
          contentReleaseId: 'mock-content-release-id',
          learningActivityId: 'mock-learning-activity-id',
          lessonUuid: 'mock-lesson-uuid'
        });
    });

    it('should track product:view_shown', () => {
      expect(defaultProps.track).toHaveBeenCalledWith({
        event: 'product:view_shown',
        version: 1,
        payload: {
          /* eslint-disable camelcase */
          view_name: 'lesson',
          locale: 'en',
          learn_language_alpha3: 'DEU'
          /* eslint-enable camelcase */
        }
      });
    });

    it('should not fetch experiment bucket', () => {
      expect(defaultProps.fetchExperimentBucket).not.toHaveBeenCalled();
    });

    it('should fetch data for smart surfaces functionality', () => {
      expect(defaultProps.fetchSmartSurfacesData).toHaveBeenCalled();
    });

    test('fetches account information', () => {
      expect(defaultProps.fetchAccount).toHaveBeenCalled();
    });

    it('should not fetch data for smart surfaces functionality when conditions are not met', () => {
      defaultProps.fetchSmartSurfacesData.mockReset();
      isAllowedReferAFriend.mockReturnValue(false);
      shallow(<LessonApp {...defaultProps} />);
      expect(defaultProps.fetchSmartSurfacesData).not.toHaveBeenCalled();
    });
  });

  describe('#fetchSmartSurfacesData', () => {
    let store;
    const getActionCreatorMock = (payload) => () => (
      Promise.resolve({
        action: {
          type: 'MOCK_ACTION',
          payload
        },
        value: payload
      })
    );

    beforeEach(() => {
      fetchTrainerItemsStatistics.mockReturnValue(getActionCreatorMock({}));
      fetchProfile.mockReturnValue(getActionCreatorMock({}));
      fetchSubscriptions.mockReturnValue(getActionCreatorMock([]));
      store = mockStore(storeMock);
    });

    it('should fetch trainer items statistics', async () => {
      await store.dispatch(fetchSmartSurfacesData());
      expect(fetchTrainerItemsStatistics).toHaveBeenCalled();
    });

    it('should notify rollbar with a warning and not interrupt UX in event of an error', async () => {
      fetchTrainerItemsStatistics.mockImplementation(() => () => Promise.reject(Error));
      await store.dispatch(fetchSmartSurfacesData());
      expect(rollbar.warning).toHaveBeenCalled();
    });

    describe('user has completed a lesson', () => {
      beforeEach(() => {
        fetchTrainerItemsStatistics.mockReturnValue(getActionCreatorMock({ totalVocabularyCount: 2 }));
      });

      it('should fetch user profile', async () => {
        await store.dispatch(fetchSmartSurfacesData());
        expect(fetchProfile).toHaveBeenCalled();
      });

      it('should fetch subscriptions if user is not a b2bUser', async () => {
        fetchProfile.mockReturnValue(getActionCreatorMock({ b2bUser: false }));
        await store.dispatch(fetchSmartSurfacesData());
        expect(fetchSubscriptions).toHaveBeenCalled();
      });

      it('should not fetch subscriptions if user is a b2bUser', async () => {
        fetchProfile.mockReturnValue(getActionCreatorMock({ b2bUser: true }));
        await store.dispatch(fetchSmartSurfacesData());
        expect(fetchSubscriptions).not.toHaveBeenCalled();
      });
    });

    describe('user has not completed a lesson', () => {
      beforeEach(async () => {
        fetchTrainerItemsStatistics.mockReturnValue(getActionCreatorMock({ totalVocabularyCount: 0 }));
        await store.dispatch(fetchSmartSurfacesData());
      });

      it('should not fetch user profile', () => {
        expect(fetchProfile).not.toHaveBeenCalled();
      });

      it('should not fetch subscriptions', () => {
        expect(fetchSubscriptions).not.toHaveBeenCalled();
      });
    });
  });
});
