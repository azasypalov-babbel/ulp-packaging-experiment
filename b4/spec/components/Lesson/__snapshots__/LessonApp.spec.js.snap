// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LessonApp default behaviour should render 1`] = `
<styled.div>
  <ApplicationErrorBoundary
    name="LessonApp"
  >
    <Connect(NavbarContainer) />
    <FeatureToggles />
    <Connect(MessagesContainer) />
    <TrainerOverlay />
    <main>
      <Component />
    </main>
    <Connect(BottomLayoutComponent) />
  </ApplicationErrorBoundary>
</styled.div>
`;
